README.md
# dotreg
Parse and stringify Windows registry files (`.reg`).

This crate deals with the text files used by regedit, not the binary registry hive format used by Windows internally.

## Installation
```toml
[dependencies]
dotreg = { git = "https://github.com/goto-bus-stop/dotreg" }
```

## Goals
This should become a part of a crate that does cross-platform Windows registry reads and writes. On windows, it can use the windows `Reg*` APIs. Elsewhere, it can serialize the updates and do `wine reg.exe import tempfile.reg`.

## License
[GPL-3.0](./LICENSE.md)


CHANGELOG.md
# dotreg change log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## 0.1.0
* Initial release.


src/lib.rs
//! Library for reading and writing Windows Registry files.

#![deny(future_incompatible)]
#![deny(nonstandard_style)]
#![deny(rust_2018_idioms)]
#![deny(unsafe_code)]
#![warn(missing_docs)]
#![warn(unused)]
#![allow(clippy::write_with_newline)]

mod parse;
mod stringify;

use encoding_rs::UTF_16LE;
use std::collections::HashMap;

/// A typed registry value.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RegValue {
    /// Arbitrary bytes.
    Binary(Vec<u8>),
    /// 32-bit integer.
    Dword(u32),
    /// 64-bit integer.
    Qword(u64),
    /// A string.
    String(String),
    /// A string with %VARS% that should be expanded.
    ExpandString(String),
    /// Multiple strings.
    MultiString(Vec<String>),
    /// A symbolic link to another registry key.
    Link(String),
    /// This value is being deleted.
    Delete,
}

/// A registry file format version.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RegFileVersion {
    /// The regedit file format used by Windows versions before Windows 2000.
    Win95,
    /// The regedit file format used by Windows versions from Windows 2000 on.
    Win2K,
}

/// A registry key, with a name and a bunch of values.
#[derive(Debug, Clone)]
pub struct RegKey {
    /// Name of the key.
    name: String,
    /// Values.
    values: HashMap<String, RegValue>,
}

impl RegKey {
    /// Create a new empty registry key with the given name.
    ///
    /// ## Examples
    /// ```rust
    /// use dotreg::RegKey;
    /// let key = RegKey::new(r"HKEY_CURRENT_USER\Control Panel\Mouse");
    /// assert_eq!(key.to_string(), "[HKEY_CURRENT_USER\\Control Panel\\Mouse]\r\n");
    /// ```
    #[inline]
    pub fn new(name: impl ToString) -> RegKey {
        RegKey {
            name: name.to_string(),
            values: Default::default(),
        }
    }

    /// Create a symbolic link to some other key.
    ///
    /// ## Examples
    /// ```rust
    /// use dotreg::RegKey;
    /// let key = RegKey::new_link(r"HKEY_CURRENT_USER\Software\MyKeyV2",
    /// r"HKEY_CURRENT_USER\Software\MyKeyV1");
    /// assert_eq!(key.to_string(), concat!(
    ///     "[HKEY_CURRENT_USER\\Software\\MyKeyV2]\r\n",
    ///     "\"SymbolicLinkValue\"=hex(6):48,4b,45,59,5f,43,55,52,52,45,4e,54,5f,55,53,45,52,5c,53,6f,66,74,77,61,72,65,5c,4d,79,4b,65,79,56,31",
    /// ));
    /// ```
    #[inline]
    pub fn new_link(name: impl ToString, target: &str) -> RegKey {
        let mut key = RegKey::new(name);
        key.set_link_target(target);
        key
    }

    /// Get the root value (@) of this key.
    #[inline]
    pub fn get_root_value(&self) -> Option<&RegValue> {
        self.get_value("@")
    }

    /// Get a value from this key.
    #[inline]
    pub fn get_value(&self, name: &str) -> Option<&RegValue> {
        self.values.get(name)
    }

    /// Set the root value (@) of this key.
    #[inline]
    pub fn set_root_value(&mut self, value: RegValue) {
        self.set_value("@", value);
    }

    /// Set a value in this key.
    #[inline]
    pub fn set_value(&mut self, name: &str, value: RegValue) {
        self.values.insert(name.to_string(), value);
    }

    /// Mark a value from this key for deletion.
    #[inline]
    pub fn delete_value(&mut self, name: &str) {
        self.set_value(name, RegValue::Delete);
    }

    /// Is this key a symbolic link?
    ///
    /// ## Examples
    /// ```rust
    /// use dotreg::RegKey;
    /// let key = RegKey::new_link(r"HKEY_CURRENT_USER\Software\MyKeyV2",
    /// r"HKEY_CURRENT_USER\Software\MyKeyV1");
    /// assert!(key.is_link());
    /// let key = RegKey::new(r"HKEY_CURRENT_USER\Software\MyKeyV1");
    /// assert!(!key.is_link());
    /// ```
    pub fn is_link(&self) -> bool {
        if self.values.len() != 1 {
            return false;
        }
        self.get_value("SymbolicLinkValue")
            .map(|val| match val {
                RegValue::Link(_) => true,
                _ => false,
            })
            .unwrap_or(false)
    }

    /// If this key is a symbolic link, get the path to the key it targets.
    pub fn link_target(&self) -> Option<&str> {
        self.get_value("SymbolicLinkValue")
            .and_then(|val| match val {
                RegValue::Link(path) => Some(&**path),
                _ => None,
            })
    }

    /// Set the symbolic link target of this key.
    #[inline]
    pub fn set_link_target(&mut self, target: &str) {
        self.set_value("SymbolicLinkValue", RegValue::Link(target.into()));
    }
}

impl ToString for RegKey {
    /// Serialize a registry key to the Win2K format.
    fn to_string(&self) -> String {
        let mut result = String::new();
        stringify::reg_key(self, &mut result);
        result
    }
}

/// Errors that may occur during parsing.
#[derive(Debug)]
pub enum ParseRegFileError {
    /// An error occurred while reading data.
    IoError(std::io::Error),
    /// Tried to read a string but it was not in the expected encoding (usually UTF-16).
    EncodingError,
    /// Parsing failed for…some reason. Hopefully this will be more detailed in the future!
    ParseError,
    /// The string was not fully parsed.
    TrailingData(String, RegFile),
}

/// Registry key modification types.
#[derive(Debug, Clone)]
pub enum RegKeyMod {
    /// Create or update a registry key.
    Update(RegKey),
    /// Delete a registry key.
    Delete,
}

/// A registry file containing registry key updates.
#[derive(Debug, Clone)]
pub struct RegFile {
    /// The version of the file.
    version: RegFileVersion,
    /// The keys in this file.
    keys: HashMap<String, RegKeyMod>,
}

impl Default for RegFile {
    #[inline]
    fn default() -> Self {
        Self {
            version: RegFileVersion::Win2K,
            keys: Default::default(),
        }
    }
}

impl RegFile {
    /// Get the file format version.
    #[inline]
    pub fn version(&self) -> RegFileVersion {
        self.version
    }

    /// Iterate over all registry keys.
    #[inline]
    pub fn keys(&self) -> impl Iterator<Item = &RegKeyMod> {
        self.keys.values()
    }

    /// Iterate mutably over all registry keys.
    #[inline]
    pub fn keys_mut(&mut self) -> impl Iterator<Item = &mut RegKeyMod> {
        self.keys.values_mut()
    }

    /// Add a registry key to this file.
    #[inline]
    pub fn add(&mut self, key: RegKey) {
        self.keys.insert(key.name.clone(), RegKeyMod::Update(key));
    }

    /// Delete a registry key.
    #[inline]
    pub fn delete(&mut self, name: &str) {
        self.keys.insert(name.to_string(), RegKeyMod::Delete);
    }

    /// Create a registry key in this file and return it.
    #[inline]
    pub fn create_key(&mut self, name: &str) -> &mut RegKey {
        let key = RegKey::new(name.to_string());
        self.add(key);
        self.get_key_mut(name).unwrap()
    }

    /// Get a registry key by name.
    #[inline]
    pub fn get_key(&self, name: &str) -> Option<&RegKey> {
        self.keys.get(name).and_then(|key_mod| {
            if let RegKeyMod::Update(ref key) = key_mod {
                Some(key)
            } else {
                None
            }
        })
    }

    /// Mutably get a registry key by name.
    #[inline]
    pub fn get_key_mut(&mut self, name: &str) -> Option<&mut RegKey> {
        self.keys.get_mut(name).and_then(|key_mod| {
            if let RegKeyMod::Update(key) = key_mod {
                Some(key)
            } else {
                None
            }
        })
    }

    /// Parse a registry file from a UTF-8 string.
    #[inline]
    pub fn parse(s: &str) -> Result<Self, ParseRegFileError> {
        let (remaining, regfile) = parse::reg_file(s).map_err(|_| ParseRegFileError::ParseError)?;
        if !remaining.is_empty() {
            // don't really wanna do the to_string() here but dotreg::read needs to be able to
            // return this while dropping the &str local.
            Err(ParseRegFileError::TrailingData(
                remaining.to_string(),
                regfile,
            ))
        } else {
            Ok(regfile)
        }
    }
}

impl ToString for RegFile {
    /// Serialize a registry file to the Win2K format.
    fn to_string(&self) -> String {
        let mut result = String::new();
        stringify::reg_file(self, &mut result);
        result
    }
}

/// Builder struct for registry files.
#[derive(Debug, Default, Clone)]
pub struct RegFileBuilder {
    inner: RegFile,
}
impl RegFileBuilder {
    /// Create a builder.
    #[inline]
    pub fn new() -> Self {
        Default::default()
    }

    /// Set the registry file version.
    ///
    /// ## Example
    /// ```rust
    /// use dotreg::{RegFileBuilder, RegFileVersion};
    /// // Default version is Win2K.
    /// let result = RegFileBuilder::new().build().to_string();
    /// assert_eq!(result, "Windows Registry Editor Version 5.00\r\n");
    /// let result = RegFileBuilder::new().version(RegFileVersion::Win2K).build().to_string();
    /// assert_eq!(result, "Windows Registry Editor Version 5.00\r\n");
    /// let result = RegFileBuilder::new().version(RegFileVersion::Win95).build().to_string();
    /// assert_eq!(result, "REGEDIT4\r\n");
    /// ```
    #[inline]
    pub fn version(mut self, version: RegFileVersion) -> Self {
        self.inner.version = version;
        self
    }

    /// Add a key update to the registry file.
    #[inline]
    pub fn update_key(mut self, key: RegKey) -> Self {
        self.inner.add(key);
        self
    }

    /// Add a key deletion to the registry file.
    ///
    /// ## Example
    /// ```rust
    /// use dotreg::RegFileBuilder;
    /// let result = RegFileBuilder::new()
    ///     .delete_key(r"HKEY_CURRENT_USER\Software\MyApp")
    ///     .build()
    ///     .to_string();
    /// assert_eq!(result, concat!(
    ///     "Windows Registry Editor Version 5.00\r\n",
    ///     "[-HKEY_CURRENT_USER\\Software\\MyApp]\r\n",
    /// ));
    /// ```
    #[inline]
    pub fn delete_key(mut self, path: &str) -> Self {
        self.inner.delete(path);
        self
    }

    /// Consume the builder and return the registry file.
    #[inline]
    pub fn build(self) -> RegFile {
        self.inner
    }
}

/// Read and parse a registry file from a byte stream.
pub fn read(mut input: impl std::io::Read) -> Result<RegFile, ParseRegFileError> {
    let mut bytes = vec![];
    input
        .read_to_end(&mut bytes)
        .map_err(ParseRegFileError::IoError)?;
    let (s, _, failed) = UTF_16LE.decode(&bytes);
    if failed {
        return Err(ParseRegFileError::EncodingError);
    }

    RegFile::parse(&s)
}

const HEADER_WIN95: &str = "REGEDIT4\r\n";
const HEADER_WIN2K: &str = "Windows Registry Editor Version 5.00\r\n";

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn builder() {
        RegFileBuilder::default().build().to_string();
    }
}


src/stringify.rs
use super::{RegFile, RegFileVersion, RegKey, RegKeyMod, RegValue, HEADER_WIN2K, HEADER_WIN95};
use std::fmt::Write;

fn header(version: RegFileVersion) -> &'static str {
    use RegFileVersion::*;
    match version {
        Win95 => HEADER_WIN95,
        Win2K => HEADER_WIN2K,
    }
}

fn reg_key_header(name: &str, output: &mut String) {
    write!(output, "[{}]\r\n", name).unwrap();
}

fn reg_key_delete(name: &str, output: &mut String) {
    write!(output, "[-{}]\r\n", name).unwrap();
}

fn quoted_string(s: &str, output: &mut String) {
    output.push('"');
    for c in s.chars() {
        match c {
            '\\' => output.push_str(r"\\"),
            '\t' => output.push_str(r"\t"),
            '\r' => output.push_str(r"\r"),
            '\n' => output.push_str(r"\n"),
            '"' => output.push_str(r#"\""#),
            c => output.push(c),
        }
    }
    output.push('"');
}

fn reg_value(value: &RegValue, output: &mut String) {
    match value {
        RegValue::Dword(n) => write!(output, "dword:{:08x}", n).unwrap(),
        RegValue::String(s) => quoted_string(s, output),
        RegValue::Binary(v) => {
            output.push_str("hex:");
            for byte in v {
                write!(output, "{:02x},", byte).unwrap();
            }
            if !v.is_empty() {
                output.pop();
            }
        }
        RegValue::Link(target) => {
            output.push_str("hex(6):");
            for byte in target.as_bytes() {
                write!(output, "{:02x},", byte).unwrap();
            }
            if !target.is_empty() {
                output.pop();
            }
        }
        _ => unimplemented!(),
    };
}

fn reg_value_line(name: &str, value: &RegValue, output: &mut String) {
    if name == "@" {
        output.push('@');
    } else {
        quoted_string(name, output);
    }
    output.push('=');
    reg_value(value, output);
}

fn reg_values<'a>(values: impl IntoIterator<Item = (&'a str, &'a RegValue)>, output: &mut String) {
    for (name, value) in values {
        reg_value_line(name, value, output);
    }
}

pub fn reg_key(key: &RegKey, output: &mut String) {
    reg_key_header(&key.name, output);
    reg_values(key.values.iter().map(|(n, v)| (&**n, v)), output);
}

pub fn reg_file(file: &RegFile, output: &mut String) {
    output.push_str(header(file.version));

    for (name, key_mod) in file.keys.iter() {
        match key_mod {
            RegKeyMod::Delete => reg_key_delete(name, output),
            RegKeyMod::Update(key) => reg_key(key, output),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn header_test() {
        let res = header(RegFileVersion::Win95);
        assert_eq!(res, "REGEDIT4\r\n");
        let res = header(RegFileVersion::Win2K);
        assert_eq!(res, "Windows Registry Editor Version 5.00\r\n");
    }

    #[test]
    fn reg_value_test() {
        let mut result = String::new();
        reg_value(&RegValue::Dword(1234), &mut result);
        assert_eq!(result, "dword:000004d2");
        result.clear();
        reg_value(&RegValue::String("abc".to_string()), &mut result);
        assert_eq!(result, "\"abc\"");
        result.clear();
        reg_value(
            &RegValue::String("esc\r\na\"pe\\t\this".to_string()),
            &mut result,
        );
        assert_eq!(result, r#""esc\r\na\"pe\\t\this""#);
        result.clear();
        reg_value(&RegValue::Binary(vec![0, 1, 2, 3]), &mut result);
        assert_eq!(result, "hex:00,01,02,03");
        result.clear();
    }
}


src/parse.rs
//! Parser functions for registry files.
//!
//! Many of these are argument-less functions because previously, they took `version` arguments
//! for Wine registry file support. Whether Wine hives should be supported by this crate is
//! currently unclear but I kept the function structure in place in case I want to add support for
//! Wine hives back in the future.
use crate::{
    RegFile, RegFileBuilder, RegFileVersion, RegKey, RegValue, HEADER_WIN2K, HEADER_WIN95,
};
use encoding_rs::UTF_16LE;
use nom::{
    branch::alt,
    bytes::complete::{escaped_transform, is_not, tag, take_while, take_while_m_n},
    character::complete::{crlf, hex_digit1},
    combinator::{map, map_res, opt},
    multi::{many0, separated_list},
    sequence::{delimited, pair, preceded, separated_pair, tuple},
    IResult,
};
use std::collections::HashMap;

/// Determine the version of the file.
fn header(input: &str) -> IResult<&str, RegFileVersion> {
    alt((
        map(tag(HEADER_WIN95), |_| RegFileVersion::Win95),
        map(tag(HEADER_WIN2K), |_| RegFileVersion::Win2K),
    ))(input)
}

/// Take and discard all whitespace characters.
fn skip_whitespace(input: &str) -> IResult<&str, ()> {
    let (input, _) = take_while(char::is_whitespace)(input)?;
    Ok((input, ()))
}

/// Read a registry key header: [HKEY_ROOT\KeyName]
fn reg_key_header() -> impl Fn(&str) -> IResult<&str, &str> {
    move |input: &str| {
        let (input, _) = skip_whitespace(input)?;
        let (input, key) = delimited(tag("["), is_not("]"), tag("]"))(input)?;

        let (input, _) = crlf(input)?;

        Ok((input, key))
    }
}

/// Read a string surrounded by double quotes.
fn quoted_string(input: &str) -> IResult<&str, String> {
    let middle = escaped_transform(is_not(r#"\""#), '\\', |esc: &str| {
        match esc.chars().next() {
            Some('\\') => Ok((&esc[1..], "\\")),
            Some('"') => Ok((&esc[1..], "\"")),
            Some('r') => Ok((&esc[1..], "\r")),
            Some('n') => Ok((&esc[1..], "\n")),
            _ => Ok((esc, "\\")),
        }
    });
    delimited(tag("\""), middle, tag("\""))(input)
}

/// Read a hex byte, two hexadecimal characters.
fn hex_byte(input: &str) -> IResult<&str, u8> {
    fn is_hex_digit(c: char) -> bool {
        c.is_digit(16)
    }

    map_res(take_while_m_n(2, 2, is_hex_digit), |s| {
        u8::from_str_radix(s, 16)
    })(input)
}

/// Read a value, like "dword:00000001"
fn reg_value() -> impl Fn(&str) -> IResult<&str, RegValue> {
    move |input: &str| {
        let dword = preceded(
            tag("dword:"),
            map_res(hex_digit1, |s| {
                u32::from_str_radix(s, 16).map(RegValue::Dword)
            }),
        );

        let comma_opt_newl = || pair(tag(","), opt(tuple((tag("\\"), crlf, skip_whitespace))));

        let binary = preceded(
            tag("hex:"),
            map(separated_list(comma_opt_newl(), hex_byte), RegValue::Binary),
        );

        let string = map(quoted_string, |s| RegValue::String(s.to_string()));

        fn parse_expand_string(mut bytes: Vec<u8>) -> RegValue {
            if bytes.ends_with(&[0, 0]) {
                bytes.pop();
                bytes.pop();
            }
            let (s, _, failed) = UTF_16LE.decode(&bytes);
            if failed {
                eprintln!("{:?}", bytes);
                panic!("should do something useful here")
            }

            RegValue::ExpandString(s.to_string())
        }

        let expand_string = preceded(
            tag("hex(2):"),
            // TODO switch to map_res
            map(
                separated_list(comma_opt_newl(), hex_byte),
                parse_expand_string,
            ),
        );

        /// Parse bytes into a list of strings.
        ///
        /// Strings are separated by two NUL bytes. The list end is marked by an empty string,
        /// i.e. four subsequent NUL bytes total.
        fn parse_multi_string(bytes: Vec<u8>) -> RegValue {
            let mut strings = vec![];
            let mut iter = bytes.iter();
            let mut last_index = 0;
            let mut index = 0;
            loop {
                match (iter.next(), iter.next()) {
                    (Some(0), Some(0)) => {
                        if index == last_index {
                            break;
                        }

                        let (s, _, failed) = UTF_16LE.decode(&bytes[last_index..index]);
                        if failed {
                            panic!("should do something useful here")
                        }
                        strings.push(s.to_string());
                        last_index = index + 2;
                    }
                    (Some(_), Some(_)) => {}
                    (Some(_), None) => panic!("uneven number of bytes"),
                    (None, None) => panic!("expected to find four NULs at the end"),
                    (None, Some(_)) => unreachable!(),
                }
                index += 2;
            }

            RegValue::MultiString(strings)
        }

        let multi_string = preceded(
            tag("hex(7):"),
            // TODO switch to map_res
            map(
                separated_list(comma_opt_newl(), hex_byte),
                parse_multi_string,
            ),
        );

        fn parse_link(bytes: Vec<u8>) -> RegValue {
            let (s, _, failed) = UTF_16LE.decode(&bytes);
            if failed {
                eprintln!("{:?}", bytes);
                panic!("should do something useful here")
            }

            RegValue::Link(s.to_string())
        }

        let link = preceded(
            tag("hex(6):"),
            // TODO switch to map_res
            map(separated_list(comma_opt_newl(), hex_byte), parse_link),
        );

        let delete = map(tag("-"), |_| RegValue::Delete);

        alt((
            dword,
            binary,
            expand_string,
            multi_string,
            link,
            string,
            delete,
        ))(input)
    }
}

fn reg_value_name(input: &str) -> IResult<&str, String> {
    quoted_string(input).or_else(move |_| map(tag("@"), ToString::to_string)(input))
}

fn reg_value_line() -> impl Fn(&str) -> IResult<&str, (String, RegValue)> {
    move |input: &str| {
        let (input, tuple) = separated_pair(reg_value_name, tag("="), reg_value())(input)?;
        let (input, _) = crlf(input)?;

        Ok((input, tuple))
    }
}

fn reg_values() -> impl Fn(&str) -> IResult<&str, Vec<(String, RegValue)>> {
    move |input: &str| {
        let (input, values) = many0(reg_value_line())(input)?;
        Ok((input, values))
    }
}

fn reg_key() -> impl Fn(&str) -> IResult<&str, RegKey> {
    move |input: &str| {
        let (input, name) = reg_key_header()(input)?;
        let (input, values) = reg_values()(input)?;
        let (input, _) = crlf(input)?;

        let mut values_map = HashMap::new();
        for (name, val) in values {
            values_map.insert(name.to_string(), val);
        }

        Ok((
            input,
            RegKey {
                name: name.to_string(),
                values: values_map,
            },
        ))
    }
}

pub fn reg_file(input: &str) -> IResult<&str, RegFile> {
    let builder = RegFileBuilder::default();

    let (input, version) = header(input)?;
    let mut builder = builder.version(version);

    let (input, keys) = many0(reg_key())(input)?;
    for key in keys {
        builder = builder.update_key(key);
    }

    Ok((input, builder.build()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn header_test() {
        let (_, res) = header("REGEDIT4\r\n").unwrap();
        assert_eq!(res, RegFileVersion::Win95);
        let (_, res) = header("Windows Registry Editor Version 5.00\r\n").unwrap();
        assert_eq!(res, RegFileVersion::Win2K);
    }

    #[test]
    fn reg_key_header_test() {
        let (_, res) = reg_key_header()("[HKEY_CURRENT_USER\\Software\\Microsoft\\Microsoft Games\\Age of Empires II: The Conquerors Expansion]\r\n").unwrap();
        assert_eq!(res, r"HKEY_CURRENT_USER\Software\Microsoft\Microsoft Games\Age of Empires II: The Conquerors Expansion");
    }

    #[test]
    fn reg_value_test() {
        let (_, res) = reg_value()("dword:0000000f").unwrap();
        assert_eq!(res, RegValue::Dword(0x0f));
        let (_, res) = reg_value()("\"192.168.178.116\"").unwrap();
        assert_eq!(res, RegValue::String("192.168.178.116".to_string()));
        let (_, res) = reg_value()("hex:30,00,00,80,10,00,00,00").unwrap();
        assert_eq!(
            res,
            RegValue::Binary(vec![0x30, 0x00, 0x00, 0x80, 0x10, 0x00, 0x00, 0x00]),
            "hex value"
        );
        let (_, res) = reg_value()("hex:30,00,00,80,\\\r\n  10,00,00,00").unwrap();
        assert_eq!(
            res,
            RegValue::Binary(vec![0x30, 0x00, 0x00, 0x80, 0x10, 0x00, 0x00, 0x00]),
            "multiline hex value"
        );
        let (_, res) = reg_value()("hex(2):43,00,3a,00,5c,00,50,00,72,00,6f,00,67,00,72,00,61,00,6d,00,20,00,46,\\\r\n00,69,00,6c,00,65,00,73,00,5c,00,49,00,6e,00,74,00,65,00,72,00,6e,00,65,00,\\\r\n74,00,20,00,45,00,78,00,70,00,6c,00,6f,00,72,00,65,00,72,00,5c,00,69,00,65,\\\r\n00,78,00,70,00,6c,00,6f,00,72,00,65,00,2e,00,65,00,78,00,65,00,2c,00,31,00,\\\r\n00,00").unwrap();
        assert_eq!(
            res,
            RegValue::ExpandString(
                r"C:\Program Files\Internet Explorer\iexplore.exe,1".to_string()
            )
        );
        let (_, res) = reg_value()("hex(6):5c,00,52,00,65,00,67,00,69,00,73,00,74,00,72,00,79,\\\r\n00,5c,00,4d,00,61,00,63,00,68,00,69,00,6e,00,65,00,5c,00,53,00,6f,00,66,00,\\\r\n74,00,77,00,61,00,72,00,65,00,5c,00,4d,00,69,00,63,00,72,00,6f,00,73,00,6f,\\\r\n00,66,00,74,00,5c,00,57,00,69,00,6e,00,64,00,6f,00,77,00,73,00,20,00,4e,00,\\\r\n54,00,5c,00,43,00,75,00,72,00,72,00,65,00,6e,00,74,00,56,00,65,00,72,00,73,\\\r\n00,69,00,6f,00,6e,00,5c,00,49,00,6d,00,61,00,67,00,65,00,20,00,46,00,69,00,\\\r\n6c,00,65,00,20,00,45,00,78,00,65,00,63,00,75,00,74,00,69,00,6f,00,6e,00,20,\\\r\n00,4f,00,70,00,74,00,69,00,6f,00,6e,00,73,00").unwrap();
        assert_eq!(res, RegValue::Link(r"\Registry\Machine\Software\Microsoft\Windows NT\CurrentVersion\Image File Execution Options".to_string().into()));
        let (_, res) = reg_value()("hex(7):4d,00,53,00,47,00,4f,00,54,00,48,00,49,00,43,00,\\\r\n2e,00,54,00,54,00,43,00,2c,00,4d,00,53,00,20,00,55,00,49,00,20,00,47,00,6f,\\\r\n00,74,00,68,00,69,00,63,00,00,00,4d,00,49,00,4e,00,47,00,4c,00,49,00,55,00,\\\r\n2e,00,54,00,54,00,43,00,2c,00,50,00,4d,00,69,00,6e,00,67,00,4c,00,69,00,55,\\\r\n00,00,00,53,00,49,00,4d,00,53,00,55,00,4e,00,2e,00,54,00,54,00,43,00,2c,00,\\\r\n53,00,69,00,6d,00,53,00,75,00,6e,00,00,00,47,00,55,00,4c,00,49,00,4d,00,2e,\\\r\n00,54,00,54,00,43,00,2c,00,47,00,75,00,6c,00,69,00,6d,00,00,00,00,00").unwrap();
        assert_eq!(
            res,
            RegValue::MultiString(vec![
                "MSGOTHIC.TTC,MS UI Gothic".to_string(),
                "MINGLIU.TTC,PMingLiU".to_string(),
                "SIMSUN.TTC,SimSun".to_string(),
                "GULIM.TTC,Gulim".to_string()
            ])
        );
        let (_, res) = reg_value()(r#""C:\\users\\goto-bus-stop\\Temp""#).unwrap();
        assert_eq!(
            res,
            RegValue::String(r"C:\users\goto-bus-stop\Temp".to_string()),
            "unescape values"
        );
        let (_, res) = reg_value()("-").unwrap();
        assert_eq!(res, RegValue::Delete, "deleted values");
    }

    #[test]
    fn reg_value_line_test() {
        let (_, res) = reg_value_line()("\"Spec Default\"=dword:00000000\r\n").unwrap();
        assert_eq!(res, ("Spec Default".to_string(), RegValue::Dword(0)));
        let (_, res) = reg_value_line()("\"Spectate IP\"=\"192.168.178.116\"\r\n").unwrap();
        assert_eq!(
            res,
            (
                "Spectate IP".to_string(),
                RegValue::String("192.168.178.116".to_string())
            )
        );
        let (_, res) = reg_value_line()("@=dword:00000000\r\n").unwrap();
        assert_eq!(res, ("@".to_string(), RegValue::Dword(0)));
    }

    #[test]
    fn reg_key_test() {
        let (_, key) = reg_key()("[HKEY_CURRENT_USER\\Software\\Microsoft\\Microsoft Games\\Age of Empires II: The Conquerors Expansion\\1.0\\EULA]\r\n\"FIRSTRUN\"=dword:00000001\r\n\r\n").unwrap();
        assert_eq!(key.name, r"HKEY_CURRENT_USER\Software\Microsoft\Microsoft Games\Age of Empires II: The Conquerors Expansion\1.0\EULA");
        let mut values = HashMap::new();
        values.insert("FIRSTRUN".to_string(), RegValue::Dword(1));
        assert_eq!(key.values, values);
    }

    #[test]
    fn reg_file_test() {
        use encoding_rs::UTF_16LE;
        let file = std::fs::read("test/input.reg").unwrap();
        let utf8 = UTF_16LE.decode(&file).0;
        reg_file(&utf8).unwrap();
    }
}
